File: examples/python/01_simple_quickstart.py
#!/usr/bin/env python3
"""
Arkilian Python SDK - Simple Quickstart Example

Demonstrates:
1. Connecting to an embedded Arkilian database
2. Using Python context managers (with Arkilian(...) as db)
3. Schema definition (DDL)
4. Parameterized inserts and queries
5. ACID transactions with rollback / commit
6. Clean resource disposal
"""

import os
import sys

# Automatically locate libarkilian.dylib/.so if ARKILIAN_LIB_PATH is not set
if "ARKILIAN_LIB_PATH" not in os.environ:
    candidate = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "c", "lib", "libarkilian.dylib"))
    if not os.path.exists(candidate):
        candidate = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "c", "lib", "libarkilian.so"))
    if os.path.exists(candidate):
        os.environ["ARKILIAN_LIB_PATH"] = candidate

# Embedded-only mode for quickstart (disable sidecar destination warnings)
os.environ["ARKILIAN_ENABLE_BACKUP"] = "0"

from arkilian import Arkilian

DB_PATH = "quickstart.sqlite"


def cleanup_db(path):
    for ext in ["", "-wal", "-shm"]:
        target = path + ext
        if os.path.exists(target):
            os.remove(target)


def main():
    print("=== Arkilian Python SDK: Simple Quickstart ===\n")
    cleanup_db(DB_PATH)

    print(f"[1] Opening Arkilian database at '{DB_PATH}'...")
    with Arkilian(DB_PATH) as db:
        # 1. Create table schema
        print("[2] Creating schema...")
        db.exec("""
            CREATE TABLE IF NOT EXISTS inventory (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                sku TEXT UNIQUE NOT NULL,
                name TEXT NOT NULL,
                price REAL NOT NULL,
                quantity INTEGER NOT NULL DEFAULT 0
            );
        """)
        print("    ✓ Table 'inventory' created.")

        # 2. Insert records using parameterized queries
        print("[3] Inserting sample inventory items...")
        insert_sql = "INSERT INTO inventory (sku, name, price, quantity) VALUES (?, ?, ?, ?)"
        db.run(insert_sql, ["SKU-101", "Mechanical Keyboard", 129.99, 45])
        db.run(insert_sql, ["SKU-102", "Ergonomic Mouse", 69.50, 120])
        db.run(insert_sql, ["SKU-103", "4K Ultra-Wide Monitor", 449.00, 18])
        print("    ✓ 3 inventory records inserted.")

        # 3. Query all items
        print("\n[4] Querying all inventory items:")
        items = db.all("SELECT id, sku, name, price, quantity FROM inventory ORDER BY id ASC")
        for item in items:
            print(f"    - ID {item['id']}: {item['name']} ({item['sku']}) | Price: ${item['price']:.2f} | In Stock: {item['quantity']}")

        # 4. Query single item by SKU
        print("\n[5] Querying item by SKU ('SKU-102')...")
        single = db.all("SELECT * FROM inventory WHERE sku = ?", ["SKU-102"])[0]
        print(f"    Found: {single['name']} (Stock: {single['quantity']}, Price: ${single['price']:.2f})")

        # 5. Atomic transaction demonstration
        print("\n[6] Executing atomic stock update transaction...")
        try:
            db.begin()
            db.run("UPDATE inventory SET quantity = quantity - 5 WHERE sku = ?", ["SKU-101"])
            db.run("UPDATE inventory SET quantity = quantity + 50 WHERE sku = ?", ["SKU-103"])
            db.commit()
            print("    ✓ Transaction committed successfully.")
        except Exception as e:
            db.rollback()
            print(f"    ✗ Transaction rolled back: {e}")

        updated_items = db.all("SELECT sku, name, quantity FROM inventory WHERE sku IN (?, ?)", ["SKU-101", "SKU-103"])
        print("    Updated Stock Levels:")
        for item in updated_items:
            print(f"    - {item['name']} ({item['sku']}): {item['quantity']} units")

    # Context manager automatically closes database
    print("\n[7] Database closed safely by context manager.")
    print("\n=== Quickstart Completed Successfully ===")


if __name__ == "__main__":
    main()

File: examples/php/01_simple_quickstart.php
<?php
/**
 * Arkilian PHP SDK - Simple Quickstart Example
 *
 * Demonstrates:
 * 1. Initializing an Arkilian embedded database
 * 2. Creating tables (DDL)
 * 3. Parameterized inserts and queries
 * 4. Fetching structured result arrays
 * 5. Atomic transactions (begin / commit / rollback)
 * 6. Clean database resource closure
 *
 * Run with:
 *   php -d ffi.enable=1 01_simple_quickstart.php
 */

require_once __DIR__ . '/Arkilian.php';

const DB_PATH = __DIR__ . '/quickstart.sqlite';

function cleanupDB(string $path): void {
    foreach (['', '-wal', '-shm'] as $ext) {
        $file = $path . $ext;
        if (file_exists($file)) {
            @unlink($file);
        }
    }
}

echo "=== Arkilian PHP SDK: Simple Quickstart ===\n\n";

cleanupDB(DB_PATH);

// Disable sidecar destination warning for local-only quickstart
putenv('ARKILIAN_ENABLE_BACKUP=0');

// 1. Initialize embedded database
echo "[1] Initializing Arkilian database at '" . basename(DB_PATH) . "'...\n";
$db = new Arkilian(DB_PATH);

// 2. Create schema
echo "[2] Creating schema...\n";
$db->exec("
    CREATE TABLE IF NOT EXISTS articles (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        title TEXT NOT NULL,
        author TEXT NOT NULL,
        views INTEGER NOT NULL DEFAULT 0,
        published_at DATETIME DEFAULT CURRENT_TIMESTAMP
    );
");
echo "    ✓ Table 'articles' created successfully.\n";

// 3. Parameterized inserts
echo "[3] Inserting sample articles...\n";
$insertSql = "INSERT INTO articles (title, author, views) VALUES (?, ?, ?)";

$articles = [
    ["Building Resilient APIs with Arkilian", "DevAdvocate", 1450],
    ["Continuous Database Replication to S3", "CloudArchitect", 3200],
    ["Zero Data-Loss Disaster Recovery Patterns", "SiteReliability", 2890],
];

foreach ($articles as $row) {
    $db->run($insertSql, $row);
}
echo "    ✓ " . count($articles) . " articles inserted.\n";

// 4. Query records
echo "\n[4] Querying all articles:\n";
$rows = $db->all("SELECT id, title, author, views FROM articles ORDER BY views DESC");
foreach ($rows as $article) {
    printf("    - ID %d | %-44s | Author: %-15s | Views: %d\n",
        $article['id'],
        $article['title'],
        $article['author'],
        $article['views']
    );
}

// 5. Atomic transaction demonstration
echo "\n[5] Executing atomic views-increment transaction...\n";
try {
    $db->begin();
    $db->run("UPDATE articles SET views = views + 500 WHERE author = ?", ["DevAdvocate"]);
    $db->commit();
    echo "    ✓ Transaction committed successfully.\n";
} catch (Throwable $e) {
    $db->rollback();
    echo "    ✗ Transaction rolled back: " . $e->getMessage() . "\n";
}

// 6. Verify updated view count
echo "\n[6] Verifying updated views:\n";
$updated = $db->all("SELECT title, views FROM articles WHERE author = ?", ["DevAdvocate"]);
foreach ($updated as $item) {
    printf("    - %s: %d views\n", $item['title'], $item['views']);
}

// 7. Clean closure
echo "\n[7] Closing database connection...\n";
$db->close();
cleanupDB(DB_PATH);
echo "    ✓ Database closed cleanly.\n";

echo "\n=== Quickstart Completed Successfully ===\n";
